src/transformations.jl

```diff
@@ -0,0 +1,208 @@
+### Transformations
+
+abstract type DataTransform end
+
+# apply the transform
+transform!(t::S, x::AbstractArray{T,1}) where {T<:Real, S<:DataTransform} = transform!(x, t, x)
+transform!(t::S, x::AbstractArray{T,2}) where {T<:Real, S<:DataTransform} = transform!(x, t, x)
+
+transform(t::S, x::AbstractArray{T,1}) where {T<:Real, S<:DataTransform} = transform!(similar(x), t, x)
+transform(t::S, x::AbstractArray{T,2}) where {T<:Real, S<:DataTransform} = transform!(similar(x), t, x)
+
+# reconstruct the original data from transformed values
+reconstruct!(t::S, x::AbstractArray{T,1}) where {T<:Real, S<:DataTransform} = reconstruct!(x, t, x)
+reconstruct!(t::S, x::AbstractArray{T,2}) where {T<:Real, S<:DataTransform} = reconstruct!(x, t, x)
+
+reconstruct(t::S, y::AbstractArray{T,1}) where {T<:Real, S<:DataTransform} = reconstruct!(similar(y), t, y)
+reconstruct(t::S, y::AbstractArray{T,2}) where {T<:Real, S<:DataTransform} = reconstruct!(similar(y), t, y)
+
+# Z-score transformation
+struct ZScoreTransform{T<:Real} <: DataTransform
+    dim::Int
+    mean::Vector{T}
+    scale::Vector{T}
+
+    function ZScoreTransform(d::Int, m::Vector{T}, s::Vector{T}) where T
+        lenm = length(m)
+        lens = length(s)
+        lenm == d || lenm == 0 || throw(DimensionMismatch("Inconsistent dimensions."))
+        lens == d || lens == 0 || throw(DimensionMismatch("Inconsistent dimensions."))
+        new{T}(d, m, s)
+    end
+end
+
+function Base.getproperty(t::ZScoreTransform, p::Symbol)
+    if p == :indim || p == :outdim
+        return t.dim
+    else
+        return getfield(t, p)
+    end
+end
+
+# fit a z-score transform
+function fit(::Type{ZScoreTransform}, X::AbstractArray{T,2}; center::Bool=true, scale::Bool=true) where T<:Real
```

Comment:
```suggestion
function fit(::Type{ZScoreTransform}, X::AbstractArray{<:Real,2}; center::Bool=true, scale::Bool=true)
```

Same below.

Reply:
I need type parameter later in the function body.

Reply:
OK, but my remark was more general.